### PR TITLE
Implement register command

### PR DESCRIPTION
**Pull request by hrsh7th**
I've implemented custom command register interface.

It will be useful when use rust-analyzer, eclipse.jdt.ls or some server that returns custom command.

For example, rust-analyzer returns `rust-analyzer.applySourceChange` command as CodeAction response and expects to handle it in client-side.

I think, it will be good to support custom command in vim-lsp-settings maybe.

#### TODO
- [ ] ~~Check capability~~
- [x] Check servers (typescript, gopls... etc)

EDIT: (mattn)

This closes #693, and related on #505,

### CODE REVIEW

File: autoload/lsp.vim [hunk 1]
```diff
@@ -166,6 +166,16 @@ function! lsp#register_server(server_info) abort
...
 endfunction
 
+"
+" lsp#register_command
+"
+" @param {command_name} = string
+" @param {callback} = funcref
+"
+function! lsp#register_command(command_name, callback) abort
+    call lsp#ui#vim#execute_command#_register(a:command_name, a:callback)
```
Comment (hrsh7th) [inline autoload/lsp.vim]: Expose api as public.

**Comment by hrsh7th** [comment]
I tested it on rust-analyzer and below codes.

```viml
    function! s:rust_analyzer_apply_source_change(context)
      let l:command = get(a:context, 'command', {})
      let l:workspace_edit = get(l:command.arguments[0], 'workspaceEdit', {})
      if !empty(l:workspace_edit)
        call lsp#utils#workspace_edit#apply_workspace_edit(l:workspace_edit)
      endif

      let l:cursor_position = get(l:command.arguments[0], 'cursorPosition', {})
      if !empty(l:cursor_position)
        call cursor(lsp#utils#positoin#_lsp_to_vim(l:command.arguments[0]['cursorPosition']))
      endif
    endfunction
    call lsp#register_command('rust-analyzer.applySourceChange', function('s:rust_analyzer_apply_source_change'))
```

**Comment by mattn** [comment]
As I updated subject of issue, this will fix issues related on client-side command doing on eclipse.jdt.js.

**Comment by hrsh7th** [comment]
coc.vim or VSCode has a similar feature.

### VSCode extension for rust-analyzer
https://github.com/rust-analyzer/rust-analyzer/blob/master/editors/code/src/main.ts#L87

### coc-prettier
https://github.com/neoclide/coc-prettier/blob/master/src/index.ts#L138

**Comment by prabirshrestha** [comment]
Sounds good and we should document with example.

**Comment by mattn** [comment]
LGTM. Please add small example and merge.

**Comment by hrsh7th** [comment]
I've updated PR.

- Display useful error message when registered command failed.
- Provide a bufnr to the command context.

I think we can better handle `capability` and some error cases.
But it is already useful and can be extended implementation so I think it is ok to merge.

**Comment by prabirshrestha** [comment]
LGTM. 

Should we make _lsp_to_vim public and also document it? same for vim_to_lsp.

File: autoload/lsp/ui/vim/execute_command.vim [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  let l:command = { 'command': l:command_name }
+  if l:command_args isnot v:null
+    let l:command['arguments'] = l:command_args
+  endif
+
+  " execute command on local.
+  if has_key(s:commands, l:command_name)
+    try
+      call s:commands[l:command_name]({
+      \   'bufnr': l:bufnr,
```
Comment (mattn) [inline autoload/lsp/ui/vim/execute_command.vim]: How about server_name? Someone may want to separate small part of common code of function.

**Comment by hrsh7th** [comment]
@prabirshrestha I've expose `lsp_to_vim` and `vim_to_lsp` as public.

File: autoload/lsp/ui/vim/execute_command.vim (hunk 2 above)
Comment (hrsh7th) [reply autoload/lsp/ui/vim/execute_command.vim]: I've added server_name as context, thanks!

**Comment by mattn** [comment]
Thank you.

**Comment by jack-michaud** [comment]
Just a heads up -- looks like 4630823 breaks using `eclipse.jdt.ls`. Upon opening any java file:
```vim
"src/main/java/PageRank/Driver.java" 43L, 1187C
Error detected while processing function <SNR>58_register_command[5]..lsp#register_comman
d[1]..lsp#ui#vim#execute_command#_register:
line    2:
E605: Exception not caught: lsp#ui#vim#execute_command#_register_command: java.apply.work
spaceEdit already registered.
Press ENTER or type command to continue
```
`java.apply.workspaceEdit` is one of those custom commands, and it runs on startup. My vimscript isn't that good, so I don't feel comfortable going in the brains to try to implement this in https://github.com/mattn/vim-lsp-settings/blob/master/settings/eclipse-jdt-ls.vim

I pinned the last working commit in my vimrc in the meantime
```
Plug 'prabirshrestha/vim-lsp', { 'commit': '4630823' }
```

**Comment by JakeElder** [comment]
Does this make it possible to use `prettier` as a formatter in the same way coc does? If it does would it be possible to get a code example?

Thank you for your work

**Pull request merged**